### PR TITLE
feat(MEN-5871): Add support for Azure Blob Storage

### PR DESCRIPTION
**Pull request by alfrunes**

### CODE REVIEW

**Comment by alfrunes** [comment]
CI depends on mendersoftware/mendertesting#218

**Comment by alfrunes** [comment]
@mender-test-bot start pipeline

File: storage/azblob/azblob.go [hunk 1]
```diff
@@ -0,0 +1,391 @@
...
+		cred, err = azblob.NewSharedKeyCredential(sk.AccountName, sk.AccountKey)
+		if err != nil {
+			return nil, err
+		}
+		var containerURI string
+		if sk.URI != nil {
+			containerURI = *sk.URI
+		} else {
+			containerURI = fmt.Sprintf(
+				"https://%s.blob.core.windows.net/%s",
```
Comment (merlin-northern) [inline storage/azblob/azblob.go:102]: magic string, could we define it?

File: storage/azblob/azblob.go [hunk 2]
```diff
@@ -0,0 +1,391 @@
...
+}
+
+type OpError struct {
+	Op      string
+	Message string
+	Reason  error
+}
+
+func (err OpError) Error() string {
+	errStr := "azblob"
```
Comment (merlin-northern) [inline storage/azblob/azblob.go:54]: ok, I do not understand the "azblob" as an error string

File: storage/azblob/azblob.go [hunk 3]
```diff
@@ -0,0 +1,391 @@
...
+	}
+	if err != nil {
+		return nil, err
+	}
+	objectStorage := &client{
+		ContainerClient: cc,
+		fileSuffix:      opt.FilenameSuffix,
+		contentType:     opt.ContentType,
+	}
+	if err := objectStorage.HealthCheck(ctx); err != nil {
```
Comment (merlin-northern) [inline storage/azblob/azblob.go:121]: is the HealthCheck also applicable to the service health check?

File: storage/azblob/azblob.go [hunk 4]
```diff
@@ -0,0 +1,391 @@
...
+	"io"
+	"net/http"
+	"net/url"
+	"path"
+	"time"
+
+	"github.com/mendersoftware/deployments/model"
+	"github.com/mendersoftware/deployments/storage"
+
+	"github.com/Azure/azure-sdk-for-go/sdk/storage/azblob"
```
Comment (merlin-northern) [inline storage/azblob/azblob.go:30]:
```suggestion
	"github.com/Azure/azure-sdk-for-go/sdk/storage/azblob"
	
	"github.com/mendersoftware/deployments/model"
	"github.com/mendersoftware/deployments/storage"
```

File: server.go [hunk 5]
```diff
@@ -89,7 +119,12 @@ func RunServer(ctx context.Context) error {
 	ds := mstore.NewDataStoreMongoWithClient(dbClient)
 
 	// Storage Layer
-	objStore, err := SetupS3(ctx)
+	var objStore storage.ObjectStorage
+	if c.GetString(dconfig.SettingDefaultStorageDefault) == "azure" {
```
Comment (merlin-northern) [inline server.go]: magic string.

File: storage/azblob/azblob.go [hunk 6]
```diff
@@ -0,0 +1,391 @@
...
+		return nil, err
+	}
+	return objectStorage, nil
+}
+
+func (c *client) HealthCheck(ctx context.Context) error {
+	_, err := c.ContainerClient.GetProperties(ctx, &azblob.ContainerGetPropertiesOptions{})
+	if err != nil {
+		return OpError{
+			Op:     "HealthCheck",
```
Comment (merlin-northern) [inline storage/azblob/azblob.go:131]: sorry for being annoying but the operations could be defined.

File: storage/azblob/azblob.go [hunk 7]
```diff
@@ -0,0 +1,391 @@
...
+	}
+	q := baseURL.Query()
+	for key, values := range qSAS {
+		for _, value := range values {
+			q.Add(key, value)
+		}
+	}
+	baseURL.RawQuery = q.Encode()
+	return baseURL.String(), nil
+}
```
Comment (merlin-northern) [inline storage/azblob/azblob.go:259]: I am missing units for this func.

File: storage/azblob/options.go [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+
+	BufferSize int
+
+	ContentType    *string
+	FilenameSuffix *string
+}
+
+func NewOptions(opts ...*Options) *Options {
+	opt := &Options{
+		BufferSize: 4 * BufferSizeMin,
```
Comment (merlin-northern) [inline storage/azblob/options.go:34]: could we get a comment in the code on why it is 4 times the size? (just in passing `BufferSizeBytes` perhaps?)

**Review by merlin-northern — COMMENTED** [review]
I left some comments, have not read till the end, but I see you are already pushing changes, so just let me know when you are done.

File: config.yaml [hunk 9]
```diff
@@ -72,13 +72,57 @@ mender-gateway: "http://mender-inventory:8080"
 
 mender-workflows: "http://mender-workflows-server:8080"
 
+
+storage:
+    # storage.default: Default storage service
+    # Must be one of ["aws", "azure"]
+    # Defaults to: "aws"
+    # Env key: DEPLOYMENTS_DEFAULT_STORAGE_SERVICE
+    default: "aws"
```
[the remark this answers is not in the document]
Comment (alfrunes) [reply config.yaml:81]: My bad, I renamed this parameter after moving it from the root to to the `storage` scope.

File: app/app.go [hunk 10]
```diff
@@ -1,16 +1,16 @@
...
-//    Unless required by applicable law or agreed to in writing, software
-//    distributed under the License is distributed on an "AS IS" BASIS,
-//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
-//    See the License for the specific language governing permissions and
-//    limitations under the License.
+//	Unless required by applicable law or agreed to in writing, software
+//	distributed under the License is distributed on an "AS IS" BASIS,
+//	WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//	See the License for the specific language governing permissions and
+//	limitations under the License.
```
[the remark this answers is not in the document]
Comment (alfrunes) [reply app/app.go:13]: It was done as part of the separate lint commit which was required to make the CI pass (I had to bump golang version to 1.18). I probably should've done this in a separate PR, but I discovered it after making this one that CI was using go1.17.

File: config/config.go [hunk 11]
```diff
@@ -160,22 +180,45 @@ func ValidateHttps(c config.Reader) error {
 	return nil
 }
 
+func ValidateStorage(c config.Reader) error {
+	svc := c.GetString(SettingDefaultStorage)
+	if svc != "aws" && svc != "azure" {
```
[the remark this answers is not in the document]
Comment (alfrunes) [reply config/config.go:185]: of course! :)

File: go.mod [hunk 12]
```diff
@@ -1,8 +1,9 @@
 module github.com/mendersoftware/deployments
 
-go 1.14
+go 1.18
```
[the remark this answers is not in the document]
Comment (alfrunes) [reply go.mod:3]: I'm using go1.19 on my local setup using auto-fomatting. I don't see any issue with formatting touched files as long as it's backward compatible changes. The Azure SDK requires go1.18, but I didn't want to go straight to 1.19 as this has some unwanted side-effects for our CI.

File: main.go [hunk 13]
```diff
@@ -85,6 +85,9 @@ func doMain(args []string) {
 	app.Before = func(args *cli.Context) error {
 
 		l := log.NewEmpty()
+		for _, alias := range dconfig.Aliases {
+			config.Config.RegisterAlias(alias.Alias, alias.Key)
+		}
```
[the remark this answers is not in the document]
Comment (alfrunes) [reply main.go:90]: These are just aliases for the options I moved out of the `aws` scope config to the more generic `storage` scope. The aliases make sure that old configurations will still work with the new configuration names.

File: model/image.go [hunk 14]
```diff
@@ -15,17 +15,28 @@
...
 )
 
+func ImagePathFromContext(ctx context.Context, id string) string {
+	imgPath := id
+	if idty := identity.FromContext(ctx); idty != nil {
+		imgPath = path.Join(idty.Tenant, id)
+	}
+	return imgPath
+}
+
```
[the remark this answers is not in the document]
Comment (alfrunes) [reply model/image.go:39]: Yes, if a path element is empty it is skipped. Also [path](https://pkg.go.dev/path) uses `/` (URL-style) as path separator.

File: storage/azblob/azblob.go (hunk 2 above)
Comment (alfrunes) [reply storage/azblob/azblob.go:54]: The error expands to `azblob( err.Op)?(: err.Message)?(: err.Reason.Error())?`.

File: storage/azblob/azblob.go (hunk 3 above)
Comment (alfrunes) [reply storage/azblob/azblob.go:121]: Yes, it is already included in the liveliness api endpoint.

Comment (alfrunes) [reply storage/azblob/azblob.go:121]: This is just a startup check to verify the credentials have access to the container/bucket (and that it exist).

**Review by kjaskiewiczz — APPROVED** [review]
it looks good, although I didn't verified if it's working, but we'll have a backend test before we'll release it, so +1; good job!

**Pull request merged**